Add false color and rotate channels in place

Arcadie opened the pull request:
Adds an attempt to colorize the image using data from both channels. Based on the algorithm by [enigmastrat](https://github.com/enigmastrat/apt137/blob/feature/false_color/channel.c#L57).

I've had some good results for images where contrast adjustment was based on telemetry:

<img width="870" alt="europe-false-color" src="https://user-images.githubusercontent.com/1472875/90427931-e8fb7d00-e0cb-11ea-9e73-f258ca7ce86d.png">

For other input files (such as `argentina.wav`), it messes things up (maybe the wav file was normalized beforehand?):

![argentina-telemetry](https://user-images.githubusercontent.com/1472875/90390197-97370080-e093-11ea-8ac7-110d7bbf2610.png)

It kinda works when using histogram equalization:

![argentina_histogram](https://user-images.githubusercontent.com/1472875/90390229-ab7afd80-e093-11ea-8771-c07a4f9d29be.png)


For the future, maybe it's a good idea to show some sliders for the ice / grass / water / dirt threshold values on the UI and let the user tweak the image if it looks wrong.

There are some things I'm not really happy about (cloning the image to get a reference to IR channel while holding a mutable reference to the image for the visible channel).  Maybe you have some suggestions as well.

Arcadie commented:
Btw, I've had great results with the wav files from here: http://www.fredvandenbosch.nl/satellites_WAV.html
I'm also curious how other peoples' images look. 😄

Arcadie commented:
The values for the width of the sync / deep space / telemetry and image data bands are from [here](http://wiki.oz9aec.net/index.php?title=File:APT_Video_Line_Format.gif):

![apt-vide-line-format.gif](http://wiki.oz9aec.net/images/e/e8/APT_Video_Line_Format.gif)

Arcadie commented:
I also noticed we can rotate the channels in place, saving the memory for a new allocation.
There are some offsets I had to adjust (1px wide), not sure where those came from, but the rotated image would not align into place otherwise, it would "bite" a piece (1-2 pixels) out of the telemetry band.

I'll try and figure out why this happens.

martinber commented:
You are the best! I'm going to look at this later with more time

> For the future, maybe it's a good idea to show some sliders for the ice / grass / water / dirt threshold values on the UI and let the user tweak the image if it looks wrong.

Yes! I don't know if I wrote that somewhere or if you read my mind

martinber commented:
Some images for reference, let me know if you want some WAV:

# Mine

## gqrx_20180917_204712_137100000

From telemetry:
![image](https://user-images.githubusercontent.com/6921727/90573972-19a3eb00-e18e-11ea-9392-77fd6f3be5c2.png)

Histogram equalization:
![image](https://user-images.githubusercontent.com/6921727/90574024-3d673100-e18e-11ea-9fe1-e2a1c0177421.png)

# gqrx_20181222_203941_137100000

From telemetry:
![image](https://user-images.githubusercontent.com/6921727/90574052-5a036900-e18e-11ea-9864-fefc96970ce3.png)

Histogram equalization:
![image](https://user-images.githubusercontent.com/6921727/90574068-64bdfe00-e18e-11ea-9688-944f00b8949a.png)

# gqrx_20190131_211722_137097000

From telemetry:
![image](https://user-images.githubusercontent.com/6921727/90574172-a189f500-e18e-11ea-83e8-4b72af30ab9a.png)

Histogram equalization:
![image](https://user-images.githubusercontent.com/6921727/90574224-bf575a00-e18e-11ea-851a-7e2ef1383a53.png)

# I don't remember where I got this from

From telemetry:
![image](https://user-images.githubusercontent.com/6921727/90574500-75bb3f00-e18f-11ea-85c1-d6095fc65a1f.png)

Histogram equalization:
![image](https://user-images.githubusercontent.com/6921727/90574749-16116380-e190-11ea-9a48-ef4a4686cb32.png)

# Provided by users to test the map overlay

## By BGNLouie

From telemetry:

Histogram equalization:

## By nosduh2

From telemetry:

Histogram equalization:

## By samarrangepas

From telemetry:

Histogram equalization:


## By KiwiEntropy

From telemetry:

Histogram equalization:


## By budude2

From telemetry:

Histogram equalization:


## By K2SDR

From telemetry:

Histogram equalization:


## By K6KZO

From telemetry:

Histogram equalization:


## By TheLAX18

From telemetry:

Histogram equalization:

martinber commented on src/noaa_apt.rs:
```diff
@@ -150,17 +149,20 @@ pub fn process(
 
     let mut img: Image = image::DynamicImage::ImageLuma8(img).into_rgba(); // convert to RGBA
 
+    if false_color {
```
I mean, right now, when histogram equalization is selected:

- First a "MinMax adjustment"  is made
- Histogram equalization is made
- False color is done

It would be better to do:

- First a "90 percent" (I don't like "From telemetry" because it won't work in short images)
- Do False color
- Do histogram equalization

But I don't know if histogram equalization can be done easily in color images. So it is enough if you leave it as it is now

martinber commented:
Seeing the output images from several WAVs, I'm seeing that this method works only in daytime passes when the surface is completely illuminated

Arcadie commented:
Hi, thanks for the feedback!

> ... But when using histogram equalization, we should do a telemetry contrast adjustment, colorize and then do histogram equalization. This is too much work, so I would add a note to the GUI saying that telemetry contrast adjustment is recommended, like you did in the command-line option

Unfortunately histogram equalization after colorization does not yield good results:

<details>
  <summary>Sample output</summary>

![argentina-hist2-color-2](https://user-images.githubusercontent.com/1472875/90637363-8ffb2880-e234-11ea-8361-1a775ef51347.png)

</details>

I expect this happens because it's a _synthetic_ image, and the histogram fills only some regions (based on the water, vegetation, clouds thresholds), and when we _equalize_ each channel separately, it leads to overblown red, blue and in-between.

I tried this on a separate branch here, if you want to take a look: https://github.com/Arcadie/noaa-apt/blob/c60c439bd108725acc570634374bb5d6367da3bd/src/processing.rs#L93

But maybe _equalization_ is not the right tool for the job, and another algorithm should be used.
Please let me know what you think.

> The constants should be specified in the settings toml file

I've added this, and also updated the documentation to include the correct path to the settings file for MacOS.

> At least three values should be editable in the GUI: the 13000, 35000 and 27000. It's not necessary to add command-line options, these users can just edit the settings toml file

Do you mind adding these to GUI? I'm struggling with the glade template and I'm unfamiliar with the available GTK widgets.
Do you edit the file manually or use something kind of IDE like https://glade.gnome.org/ ?

> 13000, 35000 and 27000.
Since we're working with u8, I'm using 50, 137 and 105 instead. (please see default settings file).

Please let me know if I've missed something.

martinber commented:
I've added three sliders to the GUI. It's working quite well, the default value for each slider is marked with a small line

![Screenshot at 2020-08-19 23-26-15](https://user-images.githubusercontent.com/6921727/90709993-d3c15280-e273-11ea-98e1-f35fd7b17b31.png)

And yes, I'm using Glade.

Also I forgot to explain better the problem with histogram equalization in color images: It can't be done for each R G B channel separately, it should be done in the "luma" channel and keep the "chroma" untouched. Or something like that, I didn't check if I'm right

I checked the code and I didn't see anything strange apart that you are changing the constants to operations and back to constants or something like that. Either way is OK for me

Also I changed the internals of how the thresholds configured in the settings propagate until they reach the `processing()` function, now there is a ColorSettings struct. I don't really like it but it was necessary to make the values modifiable from the GUI

Do you want me to merge this PR and then push my code to master? Or do you want me to create a new branch and do a PR to your Arcadie:false-color?

Arcadie commented:
That looks good!

> ... it should be done in the "luma" channel and keep the "chroma" untouched.

Oh, you're absolutely right! I should have done the RGB -> YCbCr, equalized the Y channel, and convert back to RGB. [(source)](https://stackoverflow.com/a/15009815). I'll take another stab at this one when I have some time, as it sounds promising.

> ... you are changing the constants to operations and back to constants

I reverted because I like seeing the actual value when I hover over the constant name 😄 

> Do you want me to merge this PR and then push my code to master? ...

Sure, I think this would be the easiest!

One think to consider is if the default values are "good enough" for the average quality pictures (low contrast, noise) or they need some tweaking. Either way, I'm happy to have this merged! 😃
